packages/cache/test/test-storage.js

```diff
@@ -0,0 +1,175 @@
+
+  t.is(await cache('brandName', 'barbosa'), 'barbosa');
+  t.deepEqual(Object.keys(storageNodeState), ['cache']);
+  t.deepEqual(JSON.parse(storageNodeState.cache), {
+    '{"body":"\\"brandName\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":"barbosa"}',
+      slots: [],
+    },
+  });
+
+  // One-time initialization (of 'frotz')
+  t.is(await cache('frotz', 'default'), 'default');
+  const afterFirstFrotz = {
+    '{"body":"\\"brandName\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":"barbosa"}',
+      slots: [],
+    },
+    '{"body":"\\"frotz\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":"default"}',
+      slots: [],
+    },
+  };
+  t.deepEqual(JSON.parse(storageNodeState.cache), afterFirstFrotz);
+  // no change
+  t.is(await cache('frotz', 'ignored'), 'default');
+  t.deepEqual(JSON.parse(storageNodeState.cache), afterFirstFrotz);
+
+  // cache more complex Passable
+  const complexPassable = {
+    str: 'string',
+    big: 1n,
+    num: 53,
+    arr: ['hi', 'there'],
+  };
+  t.deepEqual(
+    await cache(['complex', 'passable'], complexPassable),
+    complexPassable,
+  );
+  t.deepEqual(JSON.parse(storageNodeState.cache), {
+    ...afterFirstFrotz,
+    '{"body":"[\\"complex\\",\\"passable\\"]","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":{"arr":["hi","there"],"big":{"@qclass":"bigint","digits":"1"},"num":53,"str":"string"}}',
+      slots: [],
+    },
+  });
+});
+
+test('makeChainStorageCoordinator with remote values', async t => {
+  const { cache, storageNodeState } = setup();
+
+  const farThing = Far('farThing', { getAllegedName: () => 'dollaz' });
+
+  t.is(await cache('brand', farThing), farThing);
+  t.deepEqual(Object.keys(storageNodeState), ['cache']);
+  t.deepEqual(JSON.parse(storageNodeState.cache), {
+    '{"body":"\\"brand\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":{"@qclass":"slot","iface":"Alleged: farThing","index":0}}',
+      slots: [1],
+    },
+  });
+});
+
+test('makeChainStorageCoordinator with updater', async t => {
+  const { cache, storageNodeState } = setup();
+
+  const increment = (counter = 0) => Promise.resolve(counter + 1);
+
+  // Initial
+  t.is(await cache('counter', increment), 1);
+  t.deepEqual(Object.keys(storageNodeState), ['cache']);
+  t.deepEqual(JSON.parse(storageNodeState.cache), {
+    '{"body":"\\"counter\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":1}',
+      slots: [],
+    },
+  });
+
+  // Again without a pattern
+  t.is(await cache('counter', increment), 1);
+  t.deepEqual(Object.keys(storageNodeState), ['cache']);
+  t.deepEqual(JSON.parse(storageNodeState.cache), {
+    '{"body":"\\"counter\\"","slots":[]}': {
+      body: '{"generation":{"@qclass":"bigint","digits":"1"},"value":1}',
+      slots: [],
+    },
+  });
```

Comment:
Practically, yes. More precisely if you don't set the pattern then it's `undefined`, which will only match if the extant value is `undefined`. That will be the state if a key was set to `undefined` or not yet set.

Reply:
Ah, so is that another gap in the [README](https://github.com/Agoric/agoric-sdk/tree/6f72d00dddf3467fca0b70e1455e19f5d87adea8/packages/cache#cache-client), or is it a misimplementation of intent? [`23b94b1ebd176a7f7ae90ca64d2252ed348e8836`](https://github.com/Agoric/agoric-sdk/commit/23b94b1ebd176a7f7ae90ca64d2252ed348e8836#diff-9875c3e44f152d364645383f413b44436bf5ac1fe0c5c23e211035a6fe114eb8) inverted the behavior from its previous `guardPattern = optGuardPattern.length > 0 ? optGuardPattern[0] : M.any()`, and that looks accidental to me. Maybe @michaelfig can shed some light on this?

Reply:
> inverted the behavior... and that looks accidental to me

The original behaviour was the accident.  It makes more sense for the guard pattern to be a match on the existing value, and for the default to actually be the pattern `undefined` to match the value `undefined`, since most of the time callers want to read the current value from the cache if possible without submitting an update.

Reply:
Great, let's update the README in this PR. Suggestions:
```patch
 - the ground state for a cache key value is `undefined`.  It is impossible to distinguish a set value of `undefined` from an unset key
 - `cache(key): Promise<recentValue>` - get an eventually-consistent value for `key`
-- `cache(key, (oldValue) => ERef<newValue>): Promise<newValue>` -  call the
-  updater function transactionally with the current value of `key`, and update
-  it to `newValue`.  Rerun the updater function with a new value if the
-  transaction is stale.
-- `cache(key, (oldValue) => ERef<newValue>, guardPattern): Promise<newValue>` -
-  same as above, but only update the cache if guardPattern matches the current
-  value.  Return the updated value, or the value that matches guardPattern.
+- `cache(key, (oldValue) => ERef<newValue>, guardPattern?): Promise<newValue>` -
+  transactionally updates the value of `key` from a value matched by `guardPattern`
+  to the results of sanitizing the result of calling the updater function with
+  that value. Retries all three steps (read current value, attempt match, call updater),
+  if the transaction is stale. `guardPattern` defaults to matching only `undefined`.
+  Returns the current value after any resulting update.
+- `cache(key, passable, guardPattern?): Promise<newValue>` -
+  same as above with an updater function that ignores arguments and returns
+  `passable` (e.g., `() => passable`).
```

And noticed in passing:
```patch
 ```js
-import { makeCache, makeStoreCoordinator } from '@agoric/cache';
+import { makeCache, makeScalarStoreCoordinator } from '@agoric/cache';
 import { M } from '@agoric/store';
 import { makeScalarBigMapStore } from '@agoric/vat-data';
 
 const store = makeScalarBigMapStore('cache');
-const coordinator = makeStoreCoordinator(store);
+const coordinator = makeScalarStoreCoordinator(store);
 const cache = makeCache(coordinator);
```

Reply:
Added in d03f2df145e5b32b5aae6e77ef74c38c2ad7d7fd